src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -698,7 +757,8 @@ void ObjectSynchronizer::notify(Handle obj, TRAPS) {
   JavaThread* current = THREAD;
 
   markWord mark = obj->mark();
-  if (mark.has_locker() && current->is_lock_owned((address)mark.locker())) {
+  if ((mark.is_fast_locked() && current->lock_stack().contains(obj())) ||
+      (mark.has_locker() && current->is_lock_owned((address)mark.locker()))) {
```

Comment:
I'd prefer to see this conditionalized on UseFastlocking so that it is more obvious where the fast-locking changes are. Ditto for other places.